cpp/src/hash/sha_hash.cu

```diff
@@ -0,0 +1,793 @@
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/hashing.hpp>
+#include <cudf/detail/utilities/hash_functions.cuh>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/strings/detail/utilities.cuh>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/types.hpp>
+
+#include <hash/hash_constants.hpp>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+#include <type_traits>
+
+#include <thrust/fill.h>
+#include <thrust/for_each.h>
+#include <thrust/iterator/constant_iterator.h>
+
+namespace cudf {
+namespace detail {
+
+namespace {
+
+// SHA supported leaf data type check
+bool sha_type_check(data_type dt)
+{
+  return !is_chrono(dt) && (is_fixed_width(dt) || (dt.id() == type_id::STRING));
+}
+
+CUDA_DEVICE_CALLABLE uint32_t rotate_bits_left(uint32_t x, int8_t r)
+{
+  // This function is equivalent to (x << r) | (x >> (32 - r))
+  return __funnelshift_l(x, x, r);
+}
+
+CUDA_DEVICE_CALLABLE uint32_t rotate_bits_right(uint32_t x, int8_t r)
+{
+  // This function is equivalent to (x >> r) | (x << (32 - r))
+  return __funnelshift_r(x, x, r);
+}
+
+CUDA_DEVICE_CALLABLE uint64_t rotate_bits_right(uint64_t x, int8_t r)
+{
+  return (x >> r) | (x << (64 - r));
+}
+
+// Swap the endianness of a 32 bit value
+CUDA_DEVICE_CALLABLE uint32_t swap_endian(uint32_t x)
+{
+  // The selector 0x0123 reverses the byte order
+  return __byte_perm(x, 0, 0x0123);
+}
+
+// Swap the endianness of a 64 bit value
+// There is no CUDA intrinsic for permuting bytes in 64 bit integers
+CUDA_DEVICE_CALLABLE uint64_t swap_endian(uint64_t x)
+{
+  // Reverse the endianness of each 32 bit section
+  uint32_t low_bits  = swap_endian(static_cast<uint32_t>(x));
+  uint32_t high_bits = swap_endian(static_cast<uint32_t>(x >> 32));
+  // Reassemble a 64 bit result, swapping the low bits and high bits
+  return (static_cast<uint64_t>(low_bits) << 32) | (static_cast<uint64_t>(high_bits));
+};
+
+}  // namespace
+
+/**
+ * @brief A CRTP helper function
+ *
+ * https://www.fluentcpp.com/2017/05/19/crtp-helper/
+ *
+ * Does two things:
+ * 1. Makes "crtp" explicit in the inheritance structure of a CRTP base class.
+ * 2. Avoids having to `static_cast` in a lot of places
+ *
+ * @tparam T The derived class in a CRTP hierarchy
+ */
+template <typename T>
+struct crtp {
+  CUDA_DEVICE_CALLABLE T& underlying() { return static_cast<T&>(*this); }
+  CUDA_DEVICE_CALLABLE T const& underlying() const { return static_cast<T const&>(*this); }
+};
+
+template <typename HasherT>
+struct SHAHash : public crtp<HasherT> {
+  /**
+   * @brief Execute SHA on input data chunks.
+   *
+   * This accepts arbitrary data, handles it as bytes, and calls the hash step
+   * when the buffer is filled up to message_chunk_size bytes.
+   */
+  template <typename Hasher = HasherT>
+  void CUDA_DEVICE_CALLABLE process(uint8_t const* data,
+                                    uint32_t len,
+                                    typename Hasher::sha_intermediate_data& hash_state)
```

Comment:
@jrhemstad I took your suggestion to make the intermediate data a member variable of the Hasher classes. However, I am having trouble when I read that member variable with `auto& hash_state = this->underlying().hash_state`. The results are incorrect (as if the buffer were empty). For now, I'm allowing the Hasher to own the state and am passing in `hasher` to type-dispatch on `operator()` and providing `hasher.hash_state` by reference as a parameter to `operator()`. I confirmed that the `hasher` is created in device code. I would like to chat with you for a couple minutes when you're able - as far as I can tell, this should work. I developed some minimal test cases that work in both host and device code, and I don't see what's going wrong.

Reply:
I believe the problem is due to using CRTP incorrectly.

When using CRTP to enforce a static interface, you don't go through the derived class directly, you go through the base class template.

```
template <typename Impl>
struct SHAHash : public crtp<Impl>{
   auto process(...){
      this->underlying().hash_step(...);
   }
};

struct SHA1Hash : SHAHash<SHA1Hash> {
   auto hash_step(...){ ... }; 
};

auto foo( SHA1Hash& hash){
   hash.process(...): // this isn't right
}

template <typename Impl>
auto bar(SHAHash<Impl>& hash){
   hash.process(...); // this is what you want
}
```

Reply:
I did this and redesigned the dispatcher (per your other comment). These changes collectively fixed the problems I saw and allowed me to remove the additional template context for `Hasher = HasherT`.